File: 20221215_Thursday/20221215.go
// Leetcode: 1143. Longest Common Subsequence.
// https://leetcode.com/problems/longest-common-subsequence/
// = = = = = = = = = = = = = =
// Accepted.
// Thanks God, Jesus Christ!
// = = = = = = = = = = = = = =
// Runtime: 4 ms, faster than 78.01% of Go online submissions for Longest Common Subsequence.
// Memory Usage: 10.8 MB, less than 87.97% of Go online submissions for Longest Common Subsequence.
// 2022.12.15 Daily Challenge.
import "fmt"
var p = fmt.Println
var t1,t2 string
var dp [][]int
var m,n int

func longestCommonSubsequence(text1 string, text2 string) int {
  t1,t2 = text1,text2
  m,n = len(t1), len(t2)
  dp = make([][]int, m)
  for i:=0; i < m; i += 1 {
    dp[i] = make([]int,n)
  }
  for i := 0; i < m; i += 1 {
    for j := 0; j < n; j += 1 {
      if t1[i] == t2[j] {
        dp[i][j] = get(i-1,j-1) + 1
      } else{
        dp[i][j] = max(get(i-1,j),get(i,j-1))
      }
    }
  }
  return dp[m-1][n-1]
}

func get(i,j int) int {
  if !( 0 <= i && i < m) { return 0}
  if !( 0 <= j && j < n) { return 0}
  return dp[i][j]
}

func max(a,b int) int {
  if a > b { return a }
  return b
}

File: 20221215_Thursday/20221215.rb
# Leetcode: 1143. Longest Common Subsequence.
# https://leetcode.com/problems/longest-common-subsequence/
# = = = = = = = = = = = = = =
# Accepted.
# Thanks God, Jesus Christ!
# = = = = = = = = = = = = = =
# Runtime: 979 ms, faster than 57.58% of Ruby online submissions for Longest Common Subsequence.
# Memory Usage: 219.7 MB, less than 24.24% of Ruby online submissions for Longest Common Subsequence.
# 2022.12.15 Daily Challenge.
# @param {String} text1
# @param {String} text2
# @return {Integer}
def longest_common_subsequence(text1, text2)
  # Based on hints.
  # dp[i][j] stores maximum length of LCS for
  # t1[0..i] and t2[0..j]. Thus it has len(t1)*len(t2) elemetns.
  @dp = Array.new(text1.size).map{ Array.new(text2.size,0)}
  # if t1[i] == t2[j] => dp[i][j] = dp[i-1][j-1] + 1
  # otherwise dp[i][j] = max(dp[i-1][j],dp[i][j-1])
  @t1,@t2 = text1, text2
  (0...@t1.size).each do |i|
    (0...@t2.size).each do |j|
      if @t1[i] == @t2[j]
        @dp[i][j] = get(i-1,j-1) + 1
      else
        @dp[i][j] = [get(i-1,j),get(i,j-1)].max
      end
    end
  end
  @dp.last.last
end
def get(i,j)
  return 0 if ! i.between?(0,@dp.size - 1)
  return 0 if ! j.between?(0,@dp[0].size - 1)
  return @dp[i][j]
end
